Sources/StreamChat/AudioPlayer/AudioPlaying.swift

```diff
@@ -0,0 +1,358 @@
+        executeSeek(to: time)
+    }
+
+    // MARK: - Helpers
+
+    private func setUp() {
+        let player = self.player
+        let interval = CMTime(
+            seconds: 0.1,
+            preferredTimescale: CMTimeScale(NSEC_PER_SEC)
+        )
+
+        playerObserver.addPeriodicTimeObserver(
+            player,
+            forInterval: interval,
+            queue: nil
+        ) { [weak self] in
+            guard let self = self, self.context.isSeeking == false else {
+                return
+            }
+
+            self.updateContext { value in
+                let currentTime = player.currentTime().seconds
+                value.currentTime = currentTime.isFinite && !currentTime.isNaN
+                    ? TimeInterval(currentTime)
+                    : .zero
+
+                value.isSeeking = false
+
+                value.state = player.rate != 0 ? .playing : .paused
+
+                value.rate = .init(rawValue: player.rate)
+            }
+        }
+
+        playerObserver.addTimeControlStatusObserver(
+            player
+        ) { [weak self] newValue in
+            guard let self = self, let newValue = newValue else {
+                return
+            }
+
+            let currentPlaybackState = self.context.state
+
+            self.updateContext { value in
+                switch (newValue, currentPlaybackState) {
+                case (.playing, .playing), (.paused, .paused):
+                    break
+                case (.paused, .playing), (.paused, .loading):
+                    value.state = .paused
+                    value.rate = .zero
+                case (.playing, .paused), (.playing, .stopped), (.playing, .loading):
+                    value.state = .playing
+                default:
+                    log.debug("\(type(of: self)): No action for transition \(currentPlaybackState) -> \(newValue)", subsystems: .audioPlayback)
+                }
+            }
+        }
+
+        playerObserver.addStoppedPlaybackObserver(
+            queue: nil
+        ) { [weak self] playerItem in
+            guard
+                let self = self,
+                let playerItemURL = (playerItem.asset as? AVURLAsset)?.url,
+                let currentItemURL = (player.currentItem?.asset as? AVURLAsset)?.url,
+                playerItemURL == currentItemURL
+            else {
+                return
+            }
+            var newContext = self.context
+            newContext.state = .stopped
+            newContext.currentTime = 0
+            newContext.rate = .zero
+            newContext.isSeeking = false
+            self.context = newContext
+
+            self.delegate?.audioPlayer(self, didUpdateContext: self.context)
+        }
+    }
+
+    /// Provides thread-safe updates for the player's context and makes sure to forward any updates
+    /// to the the delegate
+    private func updateContext(
+        _ newContextProvider: (inout AudioPlaybackContext) -> Void
+    ) {
```

Comment:
I have one idea for this. It seems that it is quite important that we update the `context` through this function. But right now, nothing protects against someone manipulating the `context` property directly, which could cause issues in the future. Right now this is a bit error-prone. 

**Proposal:**
We create an `AudioPlaybackContextUpdater` component, which will have the `context` property private, and you can only change it through the `updateContext` function. WDYT?

Reply:
I like the suggestion but i'm not sure about the responsibilities. Do you mind taking a look in the `ThreadSafeValueAccessor` and let me know your thoughts?

Reply:
I would not introduce another abstraction for this now; we already have many of them in the SDK. 

But even with this new `ThreadSafeValueAccessor`, we guarantee thread-safeness. Still, we don't ensure that we use `updateContext()` every time so that you don't directly change the context to avoid unnecessary didUpdateContext events. 

As far as I understood, we use `updateContext()` here to only trigger the context's `didSet` after we change all the values we wanted, right? 

Ideally, we want to have a component that makes sure we don't forget any of these:

1. Update the context in a thread-safe manner
2. Don't forget to call the delegate when the context changes
3. Don't call the delegate more than necessary
    - Example, this should only trigger the delegate once instead of 3 times:
       ```swift
       context.state = .play
       context.rate = .half
       context.isSeeking = false
       ```

Reply:
We could still have `ThreadSafeValueAccessor` and introduce this pattern in the SDK. But it looks to me that it still makes sense to have a `AudioPlaybackContextUpdater` to make sure those 3 requirements are met. And this component can use the  `ThreadSafeValueAccessor` inside it.

Reply:
I also updated the `updateContext` and now the update to the delegate happens on didSet. Which means that there is no path you can take to update the context that is not thread safe or not updating the delegate. The `updateContext` remains as it provides an easy way to mutate (rather than having to do it inline whenever we need it to). I'll change the `ThreadSafeValueAccessor` and make it specific for the work we need it :)

Reply:
Yes, but I can still change the `context` directly without using the `updateContext` right? That is why that having only one 1 to change the context would make sense https://github.com/GetStream/stream-chat-swift/pull/2529#discussion_r1146241196

Reply:
I don't see a need for both the `AudioPlaybackContextUpdater` and `ThreadSafeValueAccessor` TBH as in the way it is now you can't make mistakes in the future. I'm trying to not introduce a "bouncing" abstraction where the player passes itself and the delegate to the update which will then receive updates and communicate them as it seems overly complicated for the size of the object.

Isn't it something that we can do in the future if needed?

Reply:
Okay makes sense, if we see issues in the future we can change it 👍